chiselFrontend/src/main/scala/chisel3/core/SeqUtils.scala

```diff
@@ -48,18 +52,70 @@ private[chisel3] object SeqUtils {
   }
 
   /** Returns the data value corresponding to the lone true predicate.
+    * This is elaborated to firrtl using a structure that should be optimized into and and/or tree.
     *
     * @note assumes exactly one true predicate, results undefined otherwise
+    *       FixedPoint values or aggregates containing FixedPoint values cause this optimized structure to be lost
     */
   def oneHotMux[T <: Data](in: Iterable[(Bool, T)]): T = macro CompileOptionsTransform.inArg
 
-  def do_oneHotMux[T <: Data](in: Iterable[(Bool, T)])(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
+  //scalastyle:off method.length cyclomatic.complexity
+  def do_oneHotMux[T <: Data](in: Iterable[(Bool, T)])
+                             (implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
     if (in.tail.isEmpty) {
       in.head._2
-    } else {
-      val masked = for ((s, i) <- in) yield Mux(s, i.asUInt, 0.U)
-      val output = cloneSupertype(in.toSeq map {_._2}, "oneHotMux")
-      output.fromBits(masked.reduceLeft(_|_))
+    }
+    else {
+      val output = cloneSupertype(in.toSeq map { _._2}, "oneHotMux")
+
+      def buildAndOrMultiplexor[TT <: Data](inputs: Iterable[(Bool, TT)]): T = {
+        val masked = for ((s, i) <- inputs) yield Mux(s, i.asUInt(), 0.U)
+        output.fromBits(masked.reduceLeft(_ | _))
+      }
+
+      output match {
+        case _: SInt =>
+          // SInt's have to be managed carefully so sign extension works
+
+          val sInts: Iterable[(Bool, SInt)] = in.collect { case (s: Bool, f: SInt) =>
+            (s, f.asTypeOf(output).asInstanceOf[SInt])
+          }
+
+          val masked = for ((s, i) <- sInts) yield Mux(s, i, 0.S)
+          output.fromBits(masked.reduceLeft(_ | _))
+
+        case _: FixedPoint =>
+          val (sels, possibleOuts) = in.toSeq.unzip
+
+          val (intWidths, binaryPoints) = in.toSeq.map { case (_, o) =>
+            val fo = o.asInstanceOf[FixedPoint]
+            require(fo.binaryPoint.known, "Mux1H requires width/binary points to be defined")
+            (fo.getWidth - fo.binaryPoint.get, fo.binaryPoint.get)
+          }.unzip
+
+          if (intWidths.distinct.length == 1 && binaryPoints.distinct.length == 1) {
+            buildAndOrMultiplexor(in)
+          }
+          else {
```

Comment:
@shunshou I tend to use ```else {``` also, so it's not all you.  I have to say I don't have strong feelings about this one, except I don't like else by itself.  ```else {``` seems to have a little more symmetry with ```if ... {``` but that's the only argument.  I'll try the other. I don't know if this can be caught by the style checker